### ISSUE TITLE
serializing learning network machines

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaAI/MLJBase.jl#844]
It doesn't seem to be possible to directly serialize `DeterministicSurrogate` machines, is this intended?  One gets
```julia
ERROR: ArgumentError: You must specify at least one operation, as in `predict=<some node>`.
Stacktrace:
 [1] machine(::DeterministicSurrogate, ::MLJBase.Source, ::Vararg{MLJBase.Source}; pair_itr::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:cache,), Tuple{Bool}}})
   @ MLJBase ~/.julia/packages/MLJBase/CtxrQ/src/composition/learning_networks/machines.jl:173
 [2] serializable(mach::Machine{DeterministicSurrogate, false})
   @ MLJBase ~/.julia/packages/MLJBase/CtxrQ/src/machines.jl:909
 [3] save(file::String, mach::Machine{DeterministicSurrogate, false})
   @ MLJBase ~/.julia/packages/MLJBase/CtxrQ/src/machines.jl:1004
 [4] top-level scope
   @ REPL[37]:1
 [5] top-level scope
   @ ~/.julia/packages/Infiltrator/qHXTS/src/Infiltrator.jl:680
```

It's coming from the constructor that is used for copying which lacks a `predict` argument.

**Comment by ablaom** [comment]
Thanks for raising this and the work at #845.

I never imagined serialization of surrogate machines. They were introduced as a way to improve on an older way of "exporting" learning networks as new standalone model types. However, I have never felt great about them, as they feel like an unnecessary abstraction. See more at #831.

At #841 I am proposing a simpler way of exporting learning networks, that gets rid of the surrogates and learning network machines altogether, and has other advantages detailed at #831. My plan was to ultimately deprecate all other export options. So, I am curious, what is your use case for using the learning network machines?

**Comment by ExpandingMan** [comment]
If I understand correctly these networks are the most general types of machine objects and are a superset of linear pipelines and ensemble models.  Most of the time when I use them, I *could* (and perhaps should) use linear pipelines instead.  In some of the more complicated cases I decide on one of multiple inner machines based on the value of a categorical value or transform outputs.  Another example is cases in which I require positive-definite output and it's far simpler to do `max.(y\hat, 0)` and the real model output is already "positive enough" that I'm not going to lose sleep over not having trained it that way.

TL;DR: My use cases usually *could* use pipeline instead.

On the other hand, I agree that these constructions can be awkward in a number of ways.  Your proposal in #841 looks good to me, though I don't think I fully understand the advantages and disadvantages of the different approaches.

Do you already support serializing the composite models there or would that remain to be added?  Having that ability would resolve this issue in my mind.

**Comment by ExpandingMan** [comment]
Ah, I see you had that in your TODO and already have it checked 👍

**Comment by ablaom** [comment]
I guess I was not clear in my question. For me, the raison d'etre for learning networks is to define new composite model types. The fact that they can be used *without* being "exported" is nice, because you can test them and play with them to test your understanding of them, but at the end of the day, they exist to define new composite model types. For example both `Pipeline` and `Stack` are implemented by exporting a learning network. 

I guess my question is: Do you have a reason for using a learning network that is *not* ultimately exported as a new type (which you will be able to serialise, and whose `report`s and `fitted_params` give you access to essential parts of the underlying component machines)?

**Comment by ExpandingMan** [comment]
> Do you have a reason for using a learning network that is not ultimately exported as a new type (which you will be able to serialise, and whose reports and fitted_params give you access to essential parts of the underlying component machines)?

No, I don't think so.  This was simply how I thought I was supposed to be doing this.  I'm not entirely sure if you are saying I was already able to do this or I *would* be able to do this after your PR.

**Comment by ablaom** [comment]
No, one has [always been able to export](https://alan-turing-institute.github.io/MLJ.jl/dev/composing_models/#Method-II:-Finer-control) learning networks as standalone composite model types.  PR #841 just makes it easier. 

I will shortly post a doc PR at MLJ which will (hopefully) make this easier for people to digest. Will copy you in for your comment. 

So I will proceed with my plan to deprecate the surrogate business, but this will be removed until the next breaking release after the coming one.  Does that sound okay?

**Comment by ExpandingMan** [comment]
Yeah, sounds great.  Now I'm just confused about what was the intended way to do this in the first place, but I'm content to wait for #841 and your updated docs, thanks.
